tests/unit_tests/core/util/buffer.cc

```diff
@@ -0,0 +1,124 @@
+ *                                                                                            //
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be       //
+ *    used to endorse or promote products derived from this software without specific         //
+ *    prior written permission.                                                               //
+ *                                                                                            //
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY        //
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    //
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL     //
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,       //
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,               //
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    //
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,          //
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF    //
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.               //
+ */
+
+#define BOOST_TEST_DYN_LINK
+
+#include <boost/test/unit_test.hpp>
+
+#include "core/util/buffer.h"
+
+struct BufferFixture
+{
+  core::Buffer<> buf;
+  std::array<std::uint8_t, 3> arr{{1, 2, 3}};
+  enum { Max = 4096 };
+};
+
+BOOST_FIXTURE_TEST_SUITE(Buffer, BufferFixture)
+
+BOOST_AUTO_TEST_CASE(Ctor)
+{
+  BOOST_CHECK_NO_THROW(core::Buffer<> buf(arr.data(), arr.size()));
+
+  core::Buffer<> buf(arr.data(), arr.size());
+  BOOST_CHECK_EQUAL(std::memcmp(buf.data(), arr.data(), buf.size()), 0);
+}
+
+BOOST_AUTO_TEST_CASE(Comparison)
+{
+  core::Buffer<> comp;
+  BOOST_CHECK(comp == buf);
+
+  comp(arr.data(), arr.size());
+  BOOST_CHECK(comp != buf);
+  BOOST_CHECK_NE(std::memcmp(comp.data(), buf.data(), comp.size()), 0);
+
+  core::Buffer<> elem(100);
+  BOOST_CHECK(comp != elem);
+  comp.clear();
+
+  BOOST_CHECK(comp != elem);
+  // New buffer should still be zero initialized
+  BOOST_CHECK_EQUAL(std::memcmp(comp.data(), elem.data(), comp.size()), 0);
+}
+
+BOOST_AUTO_TEST_CASE(Empty)
+{
+  std::array<std::uint8_t, Max> max{{}};
+  BOOST_CHECK(buf.get() == max);
+  BOOST_CHECK_EQUAL(std::memcmp(buf.data(), max.data(), max.size()), 0);
+
+  buf.clear();
+  BOOST_CHECK_EQUAL(buf.size(), 0);
+  BOOST_CHECK_EQUAL(buf.capacity(), Max);
+}
+
+BOOST_AUTO_TEST_CASE(Data)
+{
+  core::Buffer<123, 456> buf;
+  std::array<std::uint8_t, 456> data{{}};
+
+  buf(data.data(), data.size());
+  BOOST_CHECK(buf.get() == data);
+
+  data.fill(1);
+
+  BOOST_CHECK_EQUAL(buf.size(), data.size());
+  BOOST_CHECK(buf.get() != data);
+
+  buf(data.data(), data.size());
+  BOOST_CHECK(buf.get() == data);
+}
```

Comment:
I'm not sure how to catch the stack smashing error with Boost.Test, but wanted to simulate the exceptional case of an overwrite during `RouterInfo` signing.

Here is the test case that produces the error:

```diff
diff --git a/tests/unit_tests/core/util/buffer.cc b/tests/unit_tests/core/util/buffer.cc
index fbed86ef..88fa4721 100644
--- a/tests/unit_tests/core/util/buffer.cc
+++ b/tests/unit_tests/core/util/buffer.cc
@@ -97,6 +97,15 @@ BOOST_AUTO_TEST_CASE(Data)
   BOOST_CHECK(buf.get() == data);
 }
 
+BOOST_AUTO_TEST_CASE(DataOverwrite)
+{
+  std::array<std::uint8_t, Max> data{{}};
+  std::array<std::uint8_t, 64> extra{{}};
+
+  BOOST_CHECK_NO_THROW(buf(data.data(), data.size()));
+  memcpy(buf.data() + buf.size(), extra.data(), extra.size());
+}
+
```
the output from the suite:
```
Running 166 test cases...
*** stack smashing detected ***: <unknown> terminated
unknown location(0): fatal error: in "Buffer/DataOverwrite": signal: SIGABRT (application abort requested)
/kovri/tests/unit_tests/core/util/buffer.cc(100): last checkpoint: "DataOverwrite" fixture dtor

*** 1 failure is detected in the test module "Master Test Suite"
```

The same error occurs using `PrivateKeys::Sign` like in the `RouterInfo` impl.

Reply:
I can't reproduce. What compiler are you using?

Reply:
Note: none of the buildbot compilers (the ones that are online) can reproduce the issue. I can't reproduce with clang6. After testing with gcc8 though, I can reproduce.

>buf.data() + buf.size()

But what do you expect, you're telling the compiler to do a raw copy beyond the bounds of the buffer... You will get the same thing with this:

```cpp
std::array<std::uint8_t, Max> arr{{}};
std::array<std::uint8_t, Max> data{{}};
std::memcpy(arr.data() + arr.size(), data.data(), data.size());
```

Are you sure you weren't trying to do this?:

```diff
diff --git a/tests/unit_tests/core/util/buffer.cc b/tests/unit_tests/core/util/buffer.cc
index fbed86ef..911948f1 100644
--- a/tests/unit_tests/core/util/buffer.cc
+++ b/tests/unit_tests/core/util/buffer.cc
@@ -121,4 +121,11 @@ BOOST_AUTO_TEST_CASE(InvalidBuffer)
   BOOST_CHECK_THROW(core::Buffer<> buf(Max + 1), std::exception);
 }
 
+BOOST_AUTO_TEST_CASE(DataOverwrite)
+{
+  core::Buffer<0, 32> bad;
+  std::array<std::uint8_t, Max> data{{}};
+  BOOST_CHECK_THROW(bad(data.data(), data.size()), std::exception);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```

Reply:
That case shows the expected failure much better. The one I posted was to show the danger of the previous impl, and why resolving the signature TODO is important.

Resolving the signature TODO will be much safer and straightforward with this buffer util.